### PR TITLE
ENH: consistency of input args for boundaries (pd.date_range)

### LINKED ISSUE #43394: pd.date_range function not returns right bound when start=end and closed set to 'right'
```python
pd.date_range(start='2021-09-02', end='2021-09-02', freq='D', closed=None)
pd.date_range(start='2021-09-02', end='2021-09-02', freq='D', closed='right')

```

#### Problem description
the first line returns 2021-09-02
but the second line returns empty( this is incorrect )

#### Expected Output
I think if we set closed to 'right', then the date range should still returns 2021-09-02, because start date equals end date and we want the end date, the formula would looks like (2021-09-02,  2021-09-02], right date should be picked instead of returns empty

### PR DESCRIPTION
**Pull request by zyc09**
closes #43394 

xref #40245
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -1823,6 +1823,38 @@ def validate_periods(periods):
...
+        "right": (False, True),
+        "neither": (False, False),
+    }.get(inclusive)
+
+    if left_right_inclusive is None:
+        raise ValueError(
+            "Inclusive has to be either 'both', 'neither', 'left', 'right'"
+        )
+    left_inclusive, right_inclusive = left_right_inclusive
+    return left_inclusive, right_inclusive
```
Comment (attack68) [inline pandas/core/arrays/datetimelike.py:1855]: you are returning a tuple here so the individual renaming is unnecessary

File: pandas/core/indexes/datetimes.py [hunk 2]
```diff
@@ -919,6 +920,12 @@ def date_range(
     closed : {None, 'left', 'right'}, optional
         Make the interval closed with respect to the given frequency to
         the 'left', 'right', or both sides (None, the default).
+        .. deprecated:: 1.4.0
+            Argument `closed` have been deprecated
```
Comment (attack68) [inline pandas/core/indexes/datetimes.py:924]: put the comments line one space left in line with d in deprecated

File: pandas/core/arrays/datetimelike.py [hunk 3]
```diff
@@ -1823,6 +1823,37 @@ def validate_periods(periods):
     return periods
 
 
+def validate_inclusiveness(inclusive):
```
Comment (attack68) [inline pandas/core/arrays/datetimelike.py:1826]: maybe just call this validate_inclusive?

Comment (jreback) [reply pandas/core/arrays/datetimelike.py:1826]: couple of things. I would like to move this new routine & validate_entpoint to pandas/core/indexers/utils.py i think.
Also need to slightly refactor `.between_time` (which was just merged) to use the same.

Might be best to do these as a pre-cursor PR to this one.

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 4]
```diff
@@ -548,85 +548,116 @@ def test_range_closed(self, freq):
         begin = datetime(2011, 1, 1)
         end = datetime(2014, 1, 1)
 
-        closed = date_range(begin, end, closed=None, freq=freq)
-        left = date_range(begin, end, closed="left", freq=freq)
-        right = date_range(begin, end, closed="right", freq=freq)
+        both = date_range(begin, end, inclusive="both", freq=freq)
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py:551]: can you parameterize this test instead, using the `inclusive_endpoints_fixture` fixtures (its in pandas/tests/frame` so likely need to move to `pandas/conftest.py`

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 5]
```diff
@@ -548,85 +548,116 @@ def test_range_closed(self, freq):
...
     def test_range_closed_with_tz_aware_start_end(self):
         # GH12409, GH12684
         begin = Timestamp("2011/1/1", tz="US/Eastern")
         end = Timestamp("2014/1/1", tz="US/Eastern")
 
         for freq in ["1D", "3D", "2M", "7W", "3H", "A"]:
-            closed = date_range(begin, end, closed=None, freq=freq)
-            left = date_range(begin, end, closed="left", freq=freq)
-            right = date_range(begin, end, closed="right", freq=freq)
+            both = date_range(begin, end, inclusive="both", freq=freq)
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py:577]: same

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 6]
```diff
@@ -548,85 +548,116 @@ def test_range_closed(self, freq):
...
         begin = Timestamp("2011/1/1")
         end = Timestamp("2014/1/1")
         begintz = Timestamp("2011/1/1", tz="US/Eastern")
         endtz = Timestamp("2014/1/1", tz="US/Eastern")
 
         for freq in ["1D", "3D", "2M", "7W", "3H", "A"]:
-            closed = date_range(begin, end, closed=None, freq=freq, tz="US/Eastern")
-            left = date_range(begin, end, closed="left", freq=freq, tz="US/Eastern")
-            right = date_range(begin, end, closed="right", freq=freq, tz="US/Eastern")
+            both = date_range(begin, end, inclusive="both", freq=freq, tz="US/Eastern")
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py:603]: same

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 7]
```diff
@@ -548,85 +548,116 @@ def test_range_closed(self, freq):
...
+            if begintz == both[0] and endtz == both[-1]:
+                expected_neither = both[1:-1]
 
             tm.assert_index_equal(expected_left, left)
             tm.assert_index_equal(expected_right, right)
+            tm.assert_index_equal(expected_neither, neither)
 
-    @pytest.mark.parametrize("closed", ["right", "left", None])
-    def test_range_closed_boundary(self, closed):
+    @pytest.mark.parametrize("inclusive", ["right", "left", "both", "neither"])
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py:627]: same

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 8]
```diff
@@ -867,12 +909,12 @@ def test_daterange_bug_456(self):
         result = rng1.union(rng2)
         assert isinstance(result, DatetimeIndex)
 
-    @pytest.mark.parametrize("closed", ["left", "right"])
-    def test_bdays_and_open_boundaries(self, closed):
+    @pytest.mark.parametrize("inclusive", ["left", "right", "neither", "both"])
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py:912]: same

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
is this *also* closing https://github.com/pandas-dev/pandas/issues/43394 ? e.g. a bug fix, if so this needs a separate note

File: pandas/core/arrays/datetimelike.py (hunk 3 above)
Comment (attack68) [reply pandas/core/arrays/datetimelike.py:1826]: might also look at other merged prs from underlying issue that may be able to use this validator,

File: pandas/tests/indexes/datetimes/test_date_range.py (hunk 4 above)
Comment (zyc09) [reply pandas/tests/indexes/datetimes/test_date_range.py:551]: yea makes sense, will do that

**Comment by zyc09** [comment]
> is this _also_ closing #43394 ? e.g. a bug fix, if so this needs a separate note

What do you mean by separate note by the way?

**Comment by jreback** [comment]
go ahead and rebase this

File: pandas/core/arrays/datetimes.py [hunk 9]
```diff
@@ -477,10 +477,20 @@ def _generate_range(
...
 
-        if not left_closed and len(index) and index[0] == start:
-            index = index[1:]
-        if not right_closed and len(index) and index[-1] == end:
-            index = index[:-1]
+        # do not remove when one side is inclusive
+        # and removing would leave index empty
+        to_remove_any = not (
+            (left_inclusive or right_inclusive)
+            and len(index) == 1
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py:484]: shouldn't this be `len(index)` ?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimes.py (hunk 9 above)
Comment (zyc09) [reply pandas/core/arrays/datetimes.py]: mm yeah, changed that and changed the tests in test_date_range

File: pandas/conftest.py [hunk 10]
```diff
@@ -299,6 +299,11 @@ def nselect_method(request):
     return request.param
 
 
+@pytest.fixture(params=["both", "neither", "left", "right"])
```
Comment (jreback) [inline pandas/conftest.py:302]: can you put next to this one and add a doc-string that is indicative

```
@pytest.fixture(params=["left", "right", "both", "neither"])
def closed(request):
    """
    Fixture for trying all interval closed parameters.
    """
    return request.param
```

File: pandas/core/indexes/datetimes.py [hunk 11]
```diff
@@ -919,6 +920,12 @@ def date_range(
     closed : {None, 'left', 'right'}, optional
         Make the interval closed with respect to the given frequency to
         the 'left', 'right', or both sides (None, the default).
+
+        .. deprecated:: 1.4.0
+           Argument `closed` have been deprecated to standardize boundary inputs.
+           Use `inclusive` instead, to set each bound as closed or open.
+    inclusive : {"both", "neither", "left", "right"}, default "both"
+        Include boundaries; Whether to set each bound as closed or open.
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py:928]: versionadded 1.4.0

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 12]
```diff
@@ -544,89 +544,154 @@ def test_range_tz_dateutil(self):
...
-            tm.assert_index_equal(expected_right, right)
+        result_range = date_range(
+            begin, end, inclusive=inclusive_endpoints_fixture, freq=freq
+        )
+        both_range = date_range(begin, end, inclusive="both", freq=freq)
+
+        left_match = begin == both_range[0]
+        right_match = end == both_range[-1]
+
+        if inclusive_endpoints_fixture == "left" and right_match:
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py:593]: consider making this a function and share among these tests

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 13]
```diff
@@ -544,89 +544,154 @@ def test_range_tz_dateutil(self):
...
+            begin, end, inclusive="both", freq=freq, tz="US/Eastern"
+        )
 
-    @pytest.mark.parametrize("closed", ["right", "left", None])
-    def test_range_closed_boundary(self, closed):
+        left_match = begintz == both_range[0]
+        right_match = endtz == both_range[-1]
+
+        if inclusive_endpoints_fixture == "left" and right_match:
+            expected_range = both_range[:-1]
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py:634]: same (so we don't repeat ourselves)

**Review by jreback — CHANGES_REQUESTED** [review]
looking good a few comments

File: pandas/core/arrays/datetimes.py [hunk 14]
```diff
@@ -477,10 +477,20 @@ def _generate_range(
...
-            index = index[1:]
-        if not right_closed and len(index) and index[-1] == end:
-            index = index[:-1]
+        # do not remove when one side is inclusive
+        # and removing would leave index empty
+        to_remove_any = not (
+            (left_inclusive or right_inclusive)
+            and len(index)
+            and start == index[0]
+            and start == end
```
Comment (attack68) [inline pandas/core/arrays/datetimes.py:486]: when is `start != index[0]`?
when is `len(index) == 0` i.e. when will `len(index)` resolve as `False`?
when is `start == end`? 
Do any of these cases imply the others?

File: pandas/core/arrays/datetimes.py [hunk 15]
```diff
@@ -477,10 +477,20 @@ def _generate_range(
...
+        )
+
+        if to_remove_any:
+            if (not left_inclusive) and len(index) and index[0] == start:
```
Comment (attack68) [inline pandas/core/arrays/datetimes.py:490]: you are reusing the same logic tests deriving `to_remove_any`. I haven't come up with a better way but instinctively this feels a complicated logic path. Have you tried refactoring to a simpler setup?

Comment (zyc09) [reply pandas/core/arrays/datetimes.py]: I refactored the logic a little and I am not sure if it is any better though

File: pandas/tests/indexes/datetimes/test_date_range.py (hunk 12 above)
Comment (zyc09) [reply pandas/tests/indexes/datetimes/test_date_range.py]: I defined the function on the top of the test_date_range.py file, does it belong elsewhere?

File: pandas/core/indexes/datetimes.py [hunk 16]
```diff
@@ -1029,6 +1038,28 @@ def date_range(
     DatetimeIndex(['2017-01-02', '2017-01-03', '2017-01-04'],
                   dtype='datetime64[ns]', freq='D')
     """
+    if inclusive is not None and not isinstance(closed, lib.NoDefault):
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py:1041]: its possible we would want to share this (across functoions), but ok here

**Review by jreback — APPROVED** [review]
lgttm. @attack68 if comments

**Comment by jreback** [comment]
https://github.com/pandas-dev/pandas/pull/43504/checks?check_run_id=3612446736

couple of tests that need addressing (you can either assert the future warning or just change closed -> inclusive in the test)

File: pandas/core/indexes/datetimes.py [hunk 17]
```diff
@@ -919,6 +920,14 @@ def date_range(
     closed : {None, 'left', 'right'}, optional
         Make the interval closed with respect to the given frequency to
         the 'left', 'right', or both sides (None, the default).
+
+        .. deprecated:: 1.4.0
+           Argument `closed` have been deprecated to standardize boundary inputs.
```
Comment (attack68) [inline pandas/core/indexes/datetimes.py:925]: 'has' not 'have'.

File: pandas/core/indexes/datetimes.py [hunk 18]
```diff
@@ -1090,6 +1122,14 @@ def bdate_range(
     closed : str, default None
         Make the interval closed with respect to the given frequency to
         the 'left', 'right', or both sides (None).
+
+        .. deprecated:: 1.4.0
+           Argument `closed` have been deprecated to standardize boundary inputs.
```
Comment (attack68) [inline pandas/core/indexes/datetimes.py:1127]: 'has'

**Review by attack68 — APPROVED** [review]
lgtm pending green tests and some grammar corrections.

**Comment by jreback** [comment]
some comments & can you merge master. ping on green.

**Comment by attack68** [comment]
thanks @zyc09

**Pull request merged**